cartesi-rollups/tutorials/ether-wallet.md

```diff
@@ -0,0 +1,555 @@
+TEMP_DIR="temp_solc_output"
+
+# Create output and temporary directories
+mkdir -p "$TS_DIR"
+mkdir -p "$TEMP_DIR"
+
+# Function to generate ABI and export as a TypeScript variable
+generate_abi() {
+    local sol_file="$1"
+    local contract_name="$2"
+    local output_file="$TS_DIR/${contract_name}Abi.ts"
+    
+    echo "Compiling $sol_file..."
+    
+    # Compile the contract in the temporary directory
+    npx solcjs --abi "$sol_file" --base-path . --include-path node_modules/ --output-dir "$TEMP_DIR"
+    
+    # Find the generated ABI file
+    abi_file=$(find "$TEMP_DIR" -name "*_${contract_name}.abi")
+    
+    if [ ! -f "$abi_file" ]; then
+        echo "Error: ABI file not found for $contract_name"
+        return 1
+    fi
+    
+    # Read the ABI content
+    abi=$(cat "$abi_file")
+    
+    echo "Extracted ABI for $contract_name"
+    
+    # Create a TypeScript file with exported ABI
+    echo "export const ${contract_name}Abi = $abi as const;" > "$output_file"
+    
+    echo "Generated ABI for $contract_name"
+    echo "----------------------"
+}
+
+# Generate ABIs
+generate_abi "node_modules/@cartesi/rollups/contracts/dapp/CartesiDApp.sol" "CartesiDApp"
+generate_abi "node_modules/@cartesi/rollups/contracts/portals/EtherPortal.sol" "EtherPortal"
+
+# Clean up the temporary directory
+rm -rf "$TEMP_DIR"
+
+echo "ABI generation complete"
+```
+
+This script will look for all specified `.sol` files and create a TypeScript file with the ABIs in the `src/wallet/abi` directory.
+
+Now, let's make the script executable:
+
+    chmod +x generate_abis.sh
+
+
+And run it:
+
+    
+    ./generate_abis.sh
+    
+
+## Building the Ether wallet
+
+Our wallet will have two main classes: `Balance` and `Wallet`. Let's implement each of these:
+
+Create a file named `balance.ts` in the `src/wallet` directory and add the following code:
+
+```typescript
+import { Address } from "viem";
+
+export class Balance {
+  constructor(private readonly address: Address, private ether: bigint = 0n) {}
+
+  getEther(): bigint {
+    return this.ether;
+  }
+
+  increaseEther(amount: bigint): void {
+    if (amount < 0n) {
+      throw new Error(`Invalid amount: ${amount}`);
+    }
+    this.ether += amount;
+  }
+
+  decreaseEther(amount: bigint): void {
+    if (amount < 0n || this.ether < amount) {
+      throw new Error(`Invalid amount: ${amount}`);
+    }
+    this.ether -= amount;
+  }
+}
+```
+
+The `Balance` class represents an individual account's balance. It includes methods for getting, increasing, and decreasing the Ether balance.
+
+Now, create a file named `wallet.ts` in the `src/wallet` directory and add the following code:
+
+```typescript
+import {
+  Address,
+  getAddress,
+  hexToBytes,
+  stringToHex,
+  encodeFunctionData,
+} from "viem";
+import { ethers } from "ethers";
+import { Balance } from "./balance";
+import { CartesiDAppAbi } from "./abi/CartesiDAppAbi";
+import { Voucher } from "..";
```

Comment:
where is voucher coming from?

Reply:
Ok, I see it now being defined on the `index.ts`, I guess the order of things being presented might be a bit confusing.

Reply:
Haha, very familiar import convention in Danilo's projects eg. Deroll. Learning from the best 😅